doc/examples/example_symm.md

```diff
@@ -0,0 +1,206 @@
+
+Example below is similar to the previous one but uses authenticated cipher mode GCM. After encryption we take computed authentication tag and carry it along with the message for further authentication.
+
+```c++
+
+std::string message{"This is a message we going to encrypt and then, hopefully, decrypt."};
+std::vector<uint8_t> plaintext{message.begin(), message.end()};
+
+auto secretKey = utility::cryptoRandomBytes(256/8);
+
+//
+// Encryption
+//
+
+// Create cipher similarly to the previous example with two differences:
+// * use mode which supports authenticated encryption such as GCM
+// * explicitly build authenticated encryptor with `buildAuthenticatedEncryptor`.
+//
+// If you try to build authenticated cipher with mode which does not support it, `build`
+// function will throw.
+auto encryptor = AESCipherBuilder{SymmetricCipherMode::GCM, SymmetricCipherKeySize::S_256, secretKey}.buildAuthenticatedEncryptor();
+
+encryptor->update(plaintext);
+auto ciphertext = encryptor->finish();
+
+auto iv = encryptor->getIV();
+// In addition to getting IV of the encrypted block of data we also get authentication tag.
+auto authTag = encryptor->getAuthTag();
+
+//
+// Decryption
+//
+
+auto decryptor = AESCipherBuilder{SymmetricCipherMode::GCM, SymmetricCipherKeySize::S_256, secretKey}.setIV(iv).buildAuthenticatedDecryptor();
+decryptor->update(ciphertext);
+
+// Set authentication tag **before** calling `finish()`.
+decryptor->setAuthTag(authTag);
+
+std::vector<uint8_t> decryptedtext;
+try {
+    decryptedtext = decryptor->finish();
+}
+catch (const MoCOCrWException& e) {
+    std::cerr << "Decryption failed. It is possible that the message was modified after"
+    " encryption. " << e.what() << std::endl;
+}
+
+// Ensure that decrypted text matches the original message
+assert(std::equal(plaintext.begin(), plaintext.end(), decryptedtext.begin()));
+
+```
+
+## Steaming-mode
+
+Symmetric encryption interface also supports streaming use-case. It allows encryption of big chunks of data without loading them into memory. The implementation is optimized to avoid unnecessary copies of the input and output buffers when used with fixed block sizes in streaming en-/decryption.
+
+The following example demonstrates how to use stream mode with authenticated cipher.
+
+```c++
+
+std::string message;
+for (size_t i = 0; i < 1024; i++) {
+    message += "a string which will be concatenated multiple times-";
+}
+std::vector<uint8_t> plaintext{message.begin(), message.end()};
+
+auto secretKey = utility::cryptoRandomBytes(256/8);
+
+//
+// Encryption
+//
+
+auto encryptor = AESCipherBuilder{SymmetricCipherMode::GCM, SymmetricCipherKeySize::S_256, secretKey}.buildAuthenticatedEncryptor();
+
+std::vector<uint8_t> ciphertext;
+
+// Encrypt plaintext messages in 1K blocks.
+// NOTE: If `update()`-s and the following `read()` operations use the same chunk sizes you can
```

Comment:
This is actually only the case if the cipher is a stream cipher, or a block cipher and the input size produces an equivalent output size, i.e. if the input size is a multiple of the block size.

I've modified the implementation so that this also works correctly when using readAll() inside of the loop, and I think we should recommend that.

Reply:
Agree, fixed.